flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java

```diff
@@ -425,59 +425,72 @@ private void onRevokeLeadershipInternal() {
+                .forEach(
+                        contenderID ->
+                                notifyLeaderInformationChange(contenderID, leaderInformation));
     }
 
     @GuardedBy("lock")
-    private void onLeaderInformationChangeInternal(LeaderInformation leaderInformation) {
-        if (leaderContender != null) {
-            LOG.trace(
-                    "Leader node changed while {} is the leader with {}. New leader information {}.",
-                    leaderContender.getDescription(),
-                    LeaderElectionUtils.convertToString(confirmedLeaderInformation),
-                    LeaderElectionUtils.convertToString(leaderInformation));
-            if (!confirmedLeaderInformation.isEmpty()) {
-                final LeaderInformation confirmedLeaderInfo = this.confirmedLeaderInformation;
-                if (leaderInformation.isEmpty()) {
-                    LOG.debug(
-                            "Writing leader information by {} since the external storage is empty.",
-                            leaderContender.getDescription());
-                    leaderElectionDriver.publishLeaderInformation(contenderID, confirmedLeaderInfo);
-                } else if (!leaderInformation.equals(confirmedLeaderInfo)) {
-                    // the data field does not correspond to the expected leader information
-                    LOG.debug(
-                            "Correcting leader information by {}.",
-                            leaderContender.getDescription());
-                    leaderElectionDriver.publishLeaderInformation(contenderID, confirmedLeaderInfo);
-                }
-            }
-        } else {
-            LOG.debug(
-                    "Ignoring change notification since the {} has already been stopped.",
-                    leaderElectionDriver);
-        }
+    private void onLeaderInformationChangeInternal(LeaderInformationRegister leaderInformation) {
+        leaderContenderRegistry.forEach(
+                (contenderID, leaderContender) -> {
+                    final LeaderInformation newLeaderInformationForContender =
+                            leaderInformation
+                                    .forContenderID(contenderID)
+                                    .orElse(LeaderInformation.empty());
+                    final LeaderInformation confirmedLeaderInformationForContender =
+                            confirmedLeaderInformation.forContenderIDOrEmpty(contenderID);
+
+                    if (confirmedLeaderInformationForContender.isEmpty()) {
+                        LOG.trace(
+                                "Leader information changed while there's no confirmation available by the contender for contender ID '{}' for this session. Changed leader information {} will be ignored.",
+                                contenderID,
+                                LeaderElectionUtils.convertToString(
+                                        newLeaderInformationForContender));
```

Comment:
I have trouble understanding this situation. A contender got leadership, but hasn't confirmed it yet? Shouldn't we then put the leadership loss on the wire anyway? When is the contender informed about the leadership loss otherwise?

Reply:
The interface methods `onLeaderInformationChange` (for the `LeaderElectionEventHandler`) and `notifyLeaderInformationChange`/`notifyAllKnownLeaderInformation` (for the `MultipleComponentLeaderElectionDriver.Listener`) are only called if the current service has leadership acquired and informs the service that some other process updated the connection info (w/o having the right to do so). In that case, the leader has to sent its confirmed information to the HA backend to overwrite the data again. That's what's happening here.

I understand the code in a way for your case, that there is no confirmedInformation, yet, but the leadership was granted, i.e. there is a `confirmLeadership` call on the way which, as a consequence, will write the confirmed leadership to the HA backend. I guess, that this is why there was no data written to the HA backend in that case (even in the past).

But thinking about it: We might want to write the empty LeaderInformation to the HA backend in that case to avoid that monitoring services pick up the changed (false) LeaderInformation from the HA backend while the `confirmLeadership` call is delayed. :thinking:

Reply:
I also noticed that the `notifyLeaderInformationChange` method is actually implemented in a wrong way: The `LeaderInformationRegister` is used in the wrong way. I have to get back to that one. :thinking:

Reply:
> But thinking about it: We might want to write the empty LeaderInformation to the HA backend in that case to avoid that monitoring services pick up the changed (false) LeaderInformation from the HA backend while the confirmLeadership call is delayed. thinking

On the flip-side, we don't know if the leader is ready yet to accept requests yet? (because for example the dispatcher is still starting) :thinking:

Reply:
> On the flip-side, we don't know if the leader is ready yet to accept requests yet? (because for example the dispatcher is still starting) thinking
Yes, that would only be the case after the `confirmLeadership` method call finally happened that would set the `confirmedLeaderInformation` within the service and in the HA backend.

Reply:
I added the `publishLeaderInformation` call for the case you described.